### ISSUE TITLE
Separable Convolution in ASPP module is not supported according to original paper

### ISSUE THREAD

**Issue by ponjoru** [jfzhang95/pytorch-deeplab-xception#128]
I am not an expert in pytorch but... I suppose you have forgotten to add depthwise separable convolution to ASPP module. In original paper and code at the official repo (tf/models) it exists... If I am not mistaken at aspp.py following changes should be made: 
Instead of:
self.atrous_conv = nn.Conv2d(inplanes, planes, kernel_size=kernel_size,
                                            stride=1, padding=padding, dilation=dilation,, bias=False)  
Add this:
self.atrous_conv = nn.Conv2d(inplanes, planes, kernel_size=kernel_size,
                                            stride=1, padding=padding, dilation=dilation, 
                                            groups=inplanes, bias=False)  
According to the official pytorch docs: https://pytorch.org/docs/master/nn.html#conv2d
Best regards

**Comment by JeonJeongMin** [comment]
I think you are right.

Both ASPP module and Decoder don't have depthwise separable convolution in this code.
In addition to your code [Add this: ~~], pointwise convolution is needed.
For example:
self.atrous_conv = nn.Sequential(
                                     nn.Conv2d(inplanes, inplanes, kernel_size=kernel_size, stride=1, padding=padding, dilation=dilation, groups=inplanes, bias=False),
                                     nn.Conv2d(inplanes, planes, kernel_size=1, stride=1, padding=0, bias=False)
                                    )

**Comment by ponjoru** [comment]
@JeonJeongMin yeah, but there is an easier way, there is already separable convolution in Xception already implemented. So it can be used from there (but it needs a few changes). I actually moved from the deeplab. 
P.S. separable convolution didn’t make any sufficient improvement in my case

**Comment by pollackscience** [comment]
@ponjoru This is a good observation.  It should be noted, however, that depthwise separable convolution is used to improve computational efficiency and not predictive power.  In the DeepLabV3+ paper, they see basically the same predictive performance, but a large reduction in the number of Add-Multiply computations made.
